nearcore/src/cold_storage.rs

```diff
@@ -224,65 +224,88 @@ enum ColdStoreInitialMigrationResult {
-    if hot_store.get_db_kind()? != Some(near_store::metadata::DbKind::Archive)
-        || cold_store.get(DBCol::BlockMisc, HEAD_KEY)?.is_some()
-    {
-        return Ok(ColdStoreInitialMigrationResult::NoNeedForMigration);
+) -> anyhow::Result<ColdStoreMigrationResult> {
+    // Migration is only needed if cold storage is not properly initialised,
+    // i.e. if cold head is not set.
+    if cold_store.get(DBCol::BlockMisc, HEAD_KEY)?.is_some() {
+        return Ok(ColdStoreMigrationResult::NoNeedForMigration);
     }
 
-    tracing::info!(target: "cold_store", "Starting initial population of cold store");
+    tracing::info!(target: "cold_store", "Starting population of cold store.");
+    let new_cold_height = match hot_store.get_db_kind()? {
+        None => {
+            tracing::error!(target: "cold_store", "Hot store DBKind not set.");
+            return Err(anyhow::anyhow!("Hot store DBKind is not set"));
+        }
+        Some(near_store::metadata::DbKind::Hot) => {
+            tracing::info!(target: "cold_store", "Hot store DBKind is Hot.");
+            genesis_height
+        }
+        Some(near_store::metadata::DbKind::Archive) => {
+            tracing::info!(target: "cold_store", "Hot store DBKind is Archive.");
+            hot_store
+                .get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?
+                .ok_or_else(|| anyhow::anyhow!("FINAL_HEAD not found in hot storage"))?
+                .height
+        }
+        Some(kind) => {
+            tracing::error!(target: "cold_store", "Hot store DBKind not {kind:?}.");
+            return Err(anyhow::anyhow!(format!("Hot store DBKind not {kind:?}.")));
+        }
+    };
 
-    // If FINAL_HEAD is not set for hot storage something isn't right and we will probably fail in `update_cold_head`.
-    // Let's fail early.
-    let hot_final_head = hot_store
-        .get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?
-        .ok_or_else(|| anyhow::anyhow!("FINAL_HEAD not found in hot storage"))?;
-    let hot_final_head_height = hot_final_head.height;
+    tracing::info!(target: "cold_store", new_cold_height, "Determined cold storage head height after migration");
 
     let batch_size = split_storage_config.cold_store_initial_migration_batch_size;
     match copy_all_data_to_cold(cold_db.clone(), hot_store, batch_size, keep_going)? {
         CopyAllDataToColdStatus::EverythingCopied => {
-            tracing::info!(target: "cold_store", "Initial population was successful, writing cold head of height {}", hot_final_head_height);
-            update_cold_head(cold_db, hot_store, &hot_final_head_height)?;
-            Ok(ColdStoreInitialMigrationResult::SuccessfulMigration)
+            tracing::info!(target: "cold_store", new_cold_height, "Cold storage population was successful, writing cold head.");
+            update_cold_head(cold_db, hot_store, &new_cold_height)?;
```

Comment:
It seems like the new cold head is written to the db after the migration after all?